File: app-java/src/com/nfccopy/app/CardEmulationService.java
package com.nfccopy.app;

import android.content.Context;
import android.content.SharedPreferences;
import android.nfc.cardemulation.HostApduService;
import android.os.Bundle;
import org.json.JSONObject;

public class CardEmulationService extends HostApduService {

    private static final String SELECT_APDU_HEADER = "00A40400";
    private static final byte[] SW_OK = {(byte) 0x90, 0x00};
    private static final byte[] SW_NOT_FOUND = {(byte) 0x6A, (byte) 0x82};
    private static final byte[] SW_UNKNOWN = {(byte) 0x6F, 0x00};

    @Override
    public byte[] processCommandApdu(byte[] commandApdu, Bundle extras) {
        String hex = bytesToHex(commandApdu);

        SharedPreferences prefs = getSharedPreferences("nfc_cards", Context.MODE_PRIVATE);
        String activeId = prefs.getString("active_card", null);
        if (activeId == null) return SW_NOT_FOUND;

        String json = prefs.getString("card_" + activeId, null);
        if (json == null) return SW_NOT_FOUND;

        try {
            JSONObject card = new JSONObject(json);
            String uid = card.optString("uid", "");
            if (uid.isEmpty()) return SW_NOT_FOUND;

            byte[] uidBytes = hexToBytes(uid);
            byte[] response = new byte[uidBytes.length + 2];
            System.arraycopy(uidBytes, 0, response, 0, uidBytes.length);
            response[response.length - 2] = (byte) 0x90;
            response[response.length - 1] = 0x00;
            return response;
        } catch (Exception e) {
            return SW_UNKNOWN;
        }
    }

    @Override
    public void onDeactivated(int reason) {
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) sb.append(String.format("%02X", b));
        return sb.toString();
    }

    private static byte[] hexToBytes(String hex) {
        hex = hex.replace(" ", "");
        byte[] result = new byte[hex.length() / 2];
        for (int i = 0; i < result.length; i++) {
            result[i] = (byte) Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
        }
        return result;
    }
}

File: app/src/main/java/com/nfccopy/app/service/CardEmulationService.kt
package com.nfccopy.app.service

import android.nfc.cardemulation.HostApduService
import android.os.Bundle
import com.nfccopy.app.data.CardStorage

class CardEmulationService : HostApduService() {

    private lateinit var cardStorage: CardStorage

    companion object {
        private const val SELECT_APDU_HEADER = "00A40400"
        private val SW_OK = byteArrayOf(0x90.toByte(), 0x00)
        private val SW_NOT_FOUND = byteArrayOf(0x6A.toByte(), 0x82.toByte())
        private val SW_UNKNOWN = byteArrayOf(0x6F.toByte(), 0x00)
    }

    override fun onCreate() {
        super.onCreate()
        cardStorage = CardStorage(this)
    }

    override fun processCommandApdu(commandApdu: ByteArray, extras: Bundle?): ByteArray {
        val hexCommand = commandApdu.toHexString()

        // Handle SELECT command
        if (hexCommand.uppercase().startsWith(SELECT_APDU_HEADER)) {
            val activeCardId = cardStorage.getActiveCardId() ?: return SW_NOT_FOUND
            val card = cardStorage.getCard(activeCardId) ?: return SW_NOT_FOUND

            // Return UID as response data + SW_OK
            val uidBytes = card.uid.hexStringToByteArray()
            return uidBytes + SW_OK
        }

        // For GET DATA or other commands, return card raw data
        val activeCardId = cardStorage.getActiveCardId()
        if (activeCardId != null) {
            val card = cardStorage.getCard(activeCardId)
            if (card != null) {
                val responseData = card.uid.hexStringToByteArray()
                return responseData + SW_OK
            }
        }

        return SW_UNKNOWN
    }

    override fun onDeactivated(reason: Int) {
        // No-op
    }

    private fun ByteArray.toHexString(): String =
        joinToString("") { "%02X".format(it) }

    private fun String.hexStringToByteArray(): ByteArray {
        val hex = this.replace(" ", "")
        return ByteArray(hex.length / 2) { i ->
            hex.substring(i * 2, i * 2 + 2).toInt(16).toByte()
        }
    }
}
